packages/worker/src/workflow/vm-shared.ts

```diff
@@ -0,0 +1,364 @@
+export function setUnhandledRejectionHandler(getWorkflowByRunId: (runId: string) => BaseVMWorkflow | undefined): void {
+  process.on('unhandledRejection', (err, promise) => {
+    const activator = getActivator(promise);
+    const runId = activator?.info?.runId;
+    if (runId !== undefined) {
+      const workflow = getWorkflowByRunId(runId);
+      if (workflow !== undefined) {
+        workflow.setUnhandledRejection(err);
+        return;
+      }
+    }
+    // The user's logger is not accessible in this thread,
+    // dump the error information to stderr and abort.
+    console.error('Unhandled rejection', { runId }, err);
+    process.exit(1);
+  });
+}
+/**
+ * Variant of {@link cutoffStackTrace} that works with FileLocation, keep this in sync with the original implementation
+ */
+function cutoffStructuredStackTrace(stackTrace: FileLocation[]): void {
+  stackTrace.shift();
+  if (stackTrace[0].functionName === 'initAll' && stackTrace[0].filePath === 'node:internal/promise_hooks') {
+    stackTrace.shift();
+  }
+  const idx = stackTrace.findIndex(({ functionName, filePath }) => {
+    return (
+      functionName &&
+      filePath &&
+      ((/^Activator\.\S+NextHandler$/.test(functionName) &&
+        /.*[\\/]workflow[\\/](?:src|lib)[\\/]internals\.[jt]s$/.test(filePath)) ||
+        (/Script\.runInContext/.test(functionName) && /^node:vm|vm\.js$/.test(filePath)))
+    );
+  });
+  if (idx > -1) {
+    stackTrace.splice(idx);
+  }
+}
+
+function getActivator(promise: Promise<any>): Activator | undefined {
+  // Access the global scope associated with the promise (unique per workflow - vm.context)
+  // See for reference https://github.com/patriksimek/vm2/issues/32
+  const ctor = promise.constructor.constructor;
+  return ctor('return globalThis.__TEMPORAL_ACTIVATOR__')();
+}
+
+/**
+ * Internal helper to format callsite "name" portion in stack trace
+ */
+function formatCallsiteName(callsite: NodeJS.CallSite): string | null {
+  const typeName = callsite.getTypeName();
+  const methodName = callsite.getMethodName();
+  const functionName = callsite.getFunctionName();
+  const isConstructor = callsite.isConstructor();
+
+  return typeName && methodName
+    ? `${typeName}.${methodName}`
+    : isConstructor && functionName
+    ? `new ${functionName}`
+    : functionName;
+}
+
+/**
+ * Global handlers for overriding stack trace preparation and promise hooks
+ */
+export class GlobalHandlers {
+  currentStackTrace: FileLocation[] | undefined = undefined;
+  bundleFilenameToSourceMapConsumer = new Map<string, SourceMapConsumer>();
+  origPrepareStackTrace = Error.prepareStackTrace;
+  private stopPromiseHook = () => {
+    // noop
+  };
+  installed = false;
+
+  async addWorkflowBundle(workflowBundle: WorkflowBundleWithSourceMapAndFilename): Promise<void> {
+    const sourceMapConsumer = await new SourceMapConsumer(workflowBundle.sourceMap);
+    this.bundleFilenameToSourceMapConsumer.set(workflowBundle.filename, sourceMapConsumer);
```

Comment:
Is there a risk that we might register the same bundle filename multiple time? And more important... a risk that we could remove a filename multiple time? (haven't yet completed review of this class, so unsure if this can happen or not)

Reply:
The filename contain a sha sum, I don't think there's much concern here.